File: Algorithms/C++/1008-Construct_Binary_Search_Tree_from_Preorder_Traversal.cpp
// recursion
// Runtime: 8 ms, faster than 34.08% of C++ online submissions for Construct Binary Search Tree from Preorder Traversal.
// Memory Usage: 10.8 MB, less than 95.24% of C++ online submissions for Construct Binary Search Tree from Preorder Traversal.
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        return recursion(preorder, INT_MAX);
    }
private:
    int index = 0;
    TreeNode* recursion(vector<int>& preorder, int upper) {
        if (index == preorder.size() || preorder[index] > upper)
            return NULL;
        TreeNode* node = new TreeNode(preorder[index++]);
        node->left  = recursion(preorder, node->val);
        node->right = recursion(preorder, upper);
        return node;
    }
};

File: Algorithms/Python3.x/1008-Construct_Binary_Search_Tree_from_Preorder_Traversal.py
# recursion
# Runtime: 48 ms, faster than 19.18% of Python3 online submissions for Construct Binary Search Tree from Preorder Traversal.
# Memory Usage: 13.9 MB, less than 6.67% of Python3 online submissions for Construct Binary Search Tree from Preorder Traversal.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        if not preorder:
            return None
        else:
            node = TreeNode(preorder[0])
            node.left  = self.bstFromPreorder([v for v in preorder[1:] if v < preorder[0]])
            node.right = self.bstFromPreorder([v for v in preorder[1:] if v > preorder[0]])
            return node
